File: jerb/day2/part2.scala
// aoc 2020 day2 part 2
import scala.io.Source

val inputFile = "input.txt"

case class PasswordTest(
  pos1: Int,
  pos2: Int,
  char: Char,
  password: String
)


def parseLine(line: String): PasswordTest = {
  val re = """(\d+)-(\d+) ([a-zA-Z]): ([a-zA-Z]+)""".r
  line match {
    case re(p1, p2, c, pw) => PasswordTest(p1.toInt, p2.toInt, c(0), pw)
    case _ => throw new IllegalArgumentException("invalid line: " + line)
  }
}


def isValidPassword(test: PasswordTest) = 
  (test.char == test.password.charAt(test.pos1 - 1)) ^
  (test.char == test.password.charAt(test.pos2 - 1))

def solution(inputFile: String): Int =
  Source.fromFile(inputFile)
    .getLines()
    .map(parseLine)
    .filter(isValidPassword(_))
    .length

println(solution(inputFile))

File: jerb/day2/part2.py
#!/usr/bin/python3
from collections import namedtuple
import re

PasswordTest = namedtuple('PasswordTest', ('pos1', 'pos2', 'char', 'password'))


def parse_test(test):
    match = re.match(r'(\d+)-(\d+)\s+([a-zA-Z]): ([a-zA-Z]+)', test)
    if not match:
        raise ValueError('failed to parse test: ' + test)
    pos1, pos2, char, password = match.groups()
    return PasswordTest(int(pos1) - 1, int(pos2) - 1, char, password)


def is_valid_password(test):
    return test.password[test.pos1] == test.char) ^ (test.password[test.pos2] == test.char


if __name__ == '__main__':
    valid_count = 0
    with open('input.txt') as f:
        for line in f.readlines():
            test = parse_test(line.rstrip())
            if is_valid_password(test):
                valid_count += 1

    print(valid_count)
